core/src/c_api/tiledb.cc

```diff
@@ -600,146 +571,124 @@ int tiledb_array_write(
-  if (!st.ok()) {
-    strcpy(tiledb_errmsg, st.to_string().c_str());
+  if (save_error(ctx, tiledb_array->array_->write(buffers, buffer_sizes)))
     return TILEDB_ERR;
-  }
 
-  // Success
   return TILEDB_OK;
 }
 
 int tiledb_array_read(
     const TileDB_Array* tiledb_array, void** buffers, size_t* buffer_sizes) {
-  // Sanity check
-  if (!sanity_check(tiledb_array))
+  TileDB_CTX* ctx = tiledb_array->ctx_;
+
+  if (save_error(ctx, sanity_check(tiledb_array)))
     return TILEDB_ERR;
 
-  // Read
-  tiledb::Status st = tiledb_array->array_->read(buffers, buffer_sizes);
-  if (!st.ok()) {
-    strcpy(tiledb_errmsg, st.to_string().c_str());
+  if (save_error(ctx, tiledb_array->array_->read(buffers, buffer_sizes)))
     return TILEDB_ERR;
-  }
 
-  // Success
   return TILEDB_OK;
 }
 
 int tiledb_array_overflow(const TileDB_Array* tiledb_array, int attribute_id) {
-  // Sanity check
-  if (!sanity_check(tiledb_array))
+  TileDB_CTX* ctx = tiledb_array->ctx_;
+
+  if (save_error(ctx, sanity_check(tiledb_array)))
     return TILEDB_ERR;
 
   // Check overflow
   return (int)tiledb_array->array_->overflow(attribute_id);
 }
 
-int tiledb_array_consolidate(const TileDB_CTX* tiledb_ctx, const char* array) {
-  // Check array name length
-  if (array == nullptr || strlen(array) > TILEDB_NAME_MAX_LEN) {
-    std::string errmsg = "Invalid array name length";
-    PRINT_ERROR(errmsg);
-    strcpy(tiledb_errmsg, (TILEDB_ERRMSG + errmsg).c_str());
+int tiledb_array_consolidate(TileDB_CTX* ctx, const char* array) {
+  if (save_error(ctx, check_name_length("array", array, nullptr)))
     return TILEDB_ERR;
-  }
 
-  // Consolidate
-  tiledb::Status st = tiledb_ctx->storage_manager_->array_consolidate(array);
-  if (!st.ok()) {
-    strcpy(tiledb_errmsg, st.to_string().c_str());
+  if (save_error(ctx, ctx->storage_manager_->array_consolidate(array)))
     return TILEDB_ERR;
-  }
+
   return TILEDB_OK;
 }
 
 int tiledb_array_finalize(TileDB_Array* tiledb_array) {
-  // Sanity check
-  if (!sanity_check(tiledb_array) || !sanity_check(tiledb_array->tiledb_ctx_))
-    return TILEDB_ERR;
+  TileDB_CTX* ctx = tiledb_array->ctx_;
 
-  // Finalize array
-  tiledb::Status st =
-      tiledb_array->tiledb_ctx_->storage_manager_->array_finalize(
-          tiledb_array->array_);
+  if (!sanity_check(ctx).ok())
+    return TILEDB_ERR;
 
-  free(tiledb_array);
+  if (save_error(ctx, sanity_check(tiledb_array)))
+    return TILEDB_ERR;
 
-  // Error
-  if (!st.ok()) {
-    strcpy(tiledb_errmsg, st.to_string().c_str());
+  if (save_error(
+          ctx,
+          tiledb_array->ctx_->storage_manager_->array_finalize(
+              tiledb_array->array_))) {
+    free(tiledb_array);
     return TILEDB_ERR;
   }
 
-  // Success
+  free(tiledb_array);
+
   return TILEDB_OK;
 }
 
 int tiledb_array_sync(TileDB_Array* tiledb_array) {
-  // Sanity check
-  if (!sanity_check(tiledb_array) || !sanity_check(tiledb_array->tiledb_ctx_))
+  TileDB_CTX* ctx = tiledb_array->ctx_;
```

Comment:
What if `tiledb_array` is `nullptr`? I think the same holds for every function. That's the reason why I had been calling `sanity_check` on `tiledb_array` first and then again on its members.